Sources/SecureXPC/Server/XPCServer.swift

```diff
@@ -236,19 +283,64 @@ public class XPCServer {
         }
     }
     
-    private func handleMessage(connection: xpc_connection_t, message: xpc_object_t, reply: inout xpc_object_t?) throws {
-        let request = try Request(dictionary: message)
+    private func handleMessage(connection: xpc_connection_t, message: xpc_object_t) {
+        let request: Request
+        do {
+            request = try Request(dictionary: message)
+        } catch {
+            var reply = xpc_dictionary_create_reply(message)
+            self.handleError(error, expectingOtherError: false, connection: connection, reply: &reply)
+            return
+        }
+        
         guard let handler = self.routes[request.route] else {
-            throw XPCError.routeNotRegistered(request.route.pathComponents)
+            let error = XPCError.routeNotRegistered(request.route.pathComponents)
+            var reply = xpc_dictionary_create_reply(message)
+            self.handleError(error, expectingOtherError: false, connection: connection, reply: &reply)
+            return
         }
-        try handler.handle(request: request, reply: &reply)
         
-        // If a dictionary reply exists, then the message expects a reply to be sent back
-        if let reply = reply {
-            xpc_connection_send_message(connection, reply)
+        if let handler = handler as? XPCHandlerSync {
+            var reply = xpc_dictionary_create_reply(message)
+            do {
+                try handler.handle(request: request, reply: &reply)
+                
+                // If a dictionary reply exists, then the message expects a reply to be sent back
+                if let reply = reply {
+                    xpc_connection_send_message(connection, reply)
+                }
+            } catch {
+                self.handleError(error, expectingOtherError: true, connection: connection, reply: &reply)
+            }
+            
+        } else if #available(macOS 10.15.0, *), let handler = handler as? XPCHandlerAsync {
+            Task {
+                var reply = xpc_dictionary_create_reply(message)
+                do {
+                    try await handler.handle(request: request, reply: &reply)
+                    
+                    // If a dictionary reply exists, then the message expects a reply to be sent back
+                    if let reply = reply {
+                        xpc_connection_send_message(connection, reply)
+                    }
```

Comment:
Should this be moved out of this `do` block and after the catch below?

Reply:
NVM, I think my bias towards minimal `do` blocks comes from Java/Ruby/etc. where any one of the lines can be the source of the thrown exception, and it's unclear which ones you intend to catch.

I have a new-found appreciation for `try`. It makes this so clear.

Reply:
> Should this be moved out of this `do` block and after the catch below?

I know you said never mind, but FWIW it needs to be in the `do` block as it should only run if the handler succeeds. Otherwise both this logic and the logic in the `catch` block would run resulting in two sends to the same `reply` instance which would result in XPC API Misuse.